### PR TITLE
MGDAPI-4815 - Investigate adding APICast metrics to RHOAM prometheus

### PR DESCRIPTION
**Pull request by valerymo**
# Issue link
https://issues.redhat.com/browse/MGDAPI-4815
Investigate adding APICast metrics to RHOAM prometheus
Use together with 3scale-operator PR https://github.com/3scale/3scale-operator/pull/791

# What
There has been a request to have access to APICast Metrics upstream_response_time_seconds and upstream_status  and to possibly have the ability to build dashboards around these.

To make it working:
1) Current PR - ServiceMonitor `3scale-service-monitor` was added to redhat-rhoam-3scale project ( corresponding ServiceMonitor is created by reconciler in observability project)
2) 3scale PR - metrics port 9421 to be added to Services apicast-staging & apicast-production in redhat-rhoam-3scale ns (discussion with 3scale Team) 

# Verification steps
- Install Rhoam, including current PR
- edit apicast-staging Service in redhat-rhoam-3scale ns - add port:
    - name: metrics
    port: 9421
    protocol: TCP
    targetPort: 9421
_Note: adding port manually to not wait 3scale changes (this step will be not needed if 3scale provide fix)_
- Open Prometheus, and look for *upstreem* metrics. 
_Please note that metrics are hide, if no ApiCast activity._  
To try ApiCast - open 3scale UI ->Product (select API)->Integration->Configuration->Promote to stage. And run curl example
Example curl for testing:
curl "https://api-3scale-apicast-staging.apps.<your cluster>.pf40.s1.devshift.org:443/?user_key=xxxxxx"
- Check Prometheus -   upstreem* metrics should appear

### CODE REVIEW

**Comment by edodiazv** [comment]
I'm waiting this PR in order to get metrics because I can't apply these changes manually due to RHOAM namespace doesn't allow to make changes.

**Review by laurafitzgerald — COMMENTED** [review]
A couple of questions inline @valerymo

**Review by cecobask — APPROVED** [review]
Performed the verification steps and everything worked as described 👍🏻 
The `ServiceMonitor` CR was created in the 3scale product namespace and copied over to the observability product namespace. After simulating some apicast activity, the `upstream*` metrics were populated in Prometheus.

/lgtm

File: pkg/products/threescale/reconciler.go [hunk 1]
```diff
@@ -3350,3 +3356,40 @@ func (r *Reconciler) getKeycloakClientFromAccount(client k8sclient.Client, accou
 	// If the KeycloakClient wasn't found return an error
 	return nil, fmt.Errorf("failed to find the KeycloakClient for %v", accountName)
 }
+
+func (r *Reconciler) reconcileServiceMonitor(ctx context.Context, client k8sclient.Client, namespace string) (integreatlyv1alpha1.StatusPhase, error) {
```
[the remark this answers is not in the document]
Comment (valerymo) [reply pkg/products/threescale/reconciler.go:3360]: Thank you @laurafitzgerald . We discussed it yesterday, to not change, So hope it's ok (?).

File: pkg/products/threescale/reconciler.go [hunk 2]
```diff
@@ -3350,3 +3356,40 @@ func (r *Reconciler) getKeycloakClientFromAccount(client k8sclient.Client, accou
...
+		serviceMonitor.Spec = prometheus.ServiceMonitorSpec{
+			Endpoints: []prometheus.Endpoint{
+				{
+					Path: "metrics",
+					Port: "metrics",
+				},
+			},
+			Selector: metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app": "3scale-api-management",
```
[the remark this answers is not in the document]
Comment (valerymo) [reply pkg/products/threescale/reconciler.go:3383]: @laurafitzgerald , sorry, after yesterday discussion, not sure if I need to do it? But anyway, thank you for useful comment.

**Comment by laurafitzgerald** [comment]
/approve 

we discussed the creation of the serviceMonitor, in other cases we create it in the product namespace and copy it over.

**Pull request merged**